etl/steps/data/meadow/hmd/2022-12-07/life_tables.py

```diff
@@ -0,0 +1,400 @@
+
+    Returns
+    -------
+    pd.DataFrame
+        _description_
+    """
+    log.info("hmd_lt: creating dataframe from files...")
+    dfs = []
+    # Load each file
+    for f in files:
+        with open(f, "r") as f:
+            text = f.read()
+        # Get relevant fields
+        match = re.search(regex_header, text)
+        if match is not None:
+            country, _, table_str = match.group(1, 2, 3)
+        else:
+            raise ValueError(f"No match found! Please revise that source files' content matches {regex_header}.")
+        # Build country df
+        df_ = _make_df_country(country, table_str)
+        dfs.append(df_)
+    # Concatenate all country dfs
+    df = pd.concat(dfs, ignore_index=True)
+    return cast(pd.DataFrame, df)
+
+
+def _make_df_country(country: str, table: str) -> pd.DataFrame:
+    """Create dataframe for individual country."""
+    # Remove starting/ending spaces
+    table = table.strip()
+    # Remove spacing after newline
+    table = re.sub(r"\n\s+", "\n", table)
+    # Replace spacing with tabs
+    table = re.sub(r"[^\S\r\n]+", "\t", table)
+    # Build df
+    df = pd.read_csv(StringIO(table), sep="\t")
+    # Assign country
+    df = df.assign(Country=country)
+    # # Filter columns
+    # df = df[["Country", "Year", "ex"]].rename(columns={"ex": "life_expectancy"})
+    return df
+
+
+def clean_df(df: pd.DataFrame) -> pd.DataFrame:
+    """Clean dataframe.
+
+    Orders columns, renames columns, checks for missing values and sets dtypes.
+
+    Parameters
+    ----------
+    df : pd.DataFrame
+        Initial dataframe.
+
+    Returns
+    -------
+    pd.DataFrame
+        Cleaned dataframe.
+    """
+    log.info("hmd_lt: cleaning dataframe...")
+    # Order columns
+    cols_first = ["Country", "Year"]
+    cols = cols_first + [col for col in df.columns if col not in cols_first]
+    df = df[cols]
+    # Rename columns
+    df = _clean_rename_columns_df(df)
+    # Correct missing data
+    df = _clean_correct_missing_data(df)
+    # Set dtypes
+    df = _clean_set_dtypes_df(df)
+    return df
+
+
+def _clean_rename_columns_df(df: pd.DataFrame) -> pd.DataFrame:
+    """Rename columns."""
+    return df.rename(columns=COLUMNS_RENAME)
+
+
+def _clean_correct_missing_data(df: pd.DataFrame) -> pd.DataFrame:
+    """Checks on missing data."""
+    # Expected missing data
+    num_rows_missing_expected = 0.01
+    countries_missing_data_expected = 1
+    # Find missing data
+    rows_missing = df[df.central_death_rate == "."]
+    num_rows_missing = len(rows_missing) / len(df)
+    countries_missing_data = rows_missing.country.unique()
+    # Run checks
+    assert num_rows_missing < num_rows_missing_expected, (
+        f"More missing data than expected was found! {round(num_rows_missing*100, 2)} rows missing, but"
+        f" {round(num_rows_missing_expected*100,2)} were expected."
+    )
+    assert len(countries_missing_data) <= countries_missing_data_expected, (
```

Comment:
If you have identified exactly what the missing data is, a safer option would be to actually list, e.g., the country that has missing data, on the top of the script. Then assert that the data is indeed missing (in case this changes on next update), and that there is no additional data missing.

Reply:
good idea!